File: 25_oct_2019.py
"""
Question:
Given a string containing just the characters '{', '(', ')', '}', '[', and ']', determine if the string is valid
An input string is valid if -
> Open brackets are closed by the same type of bracket
> Open brackets are closed in the same order
> Note than an empty string is also considered valid
"""
import unittest

class Solution:

    def isValid(self, string):
        if len(string) == 0:
            return True

        pmap = {'{': '}', '(': ')', '[': ']'}
        stack, top = [], -1

        for char in string:
            if top < 0:
                stack.append(char)
                top += 1
                continue

            stop = stack[top]
            try:
                if char == pmap[stop]:
                    stack.pop(top)
                    top -= 1
                    continue
            except KeyError as e:
                return False

            stack.append(char)
            top += 1

        return top < 0

class TestCases(unittest.TestCase):

    def setUp(self) -> None:
        self.sol = Solution()

    def test_isValid(self):
        self.assertTrue(self.sol.isValid('()()()'))
        self.assertTrue(self.sol.isValid('({})[]'))
        self.assertTrue(self.sol.isValid('[[((){})]]'))
        self.assertTrue(self.sol.isValid(''))

        self.assertFalse(self.sol.isValid('()(()'))
        self.assertFalse(self.sol.isValid('{[}{]]{}'))
        self.assertFalse(self.sol.isValid('{]'))

if __name__ == "__main__":
    print(Solution().isValid('{[}{]]{}'))

File: 25_oct_2019.cpp
/*
Question:
Given a string containing just the characters '{', '(', ')', '}', '[', and ']', determine if the string is valid
An input string is valid if -
> Open brackets are closed by the same type of bracket
> Open brackets are closed in the same order
> Note than an empty string is also considered valid
*/

#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        map<char, char> vmap = {{'{', '}'}, {'[', ']'}, {'(', ')'}};
        char cchar;

        for (auto c: s) {
            if (stack.empty()) {
                stack.push_back(c);
                continue;
            }
            cchar = stack[stack.size()-1];
            try {
                if (c == vmap[cchar]) {
                    stack.pop_back();
                    continue;
                }
            }
            catch (...) {
                return false;
            }
            stack.push_back(c);
        }

        return (bool) stack.empty();
    }
};

int main() {
    Solution s;
    cout << s.isValid("()()()");
}